### ISSUE TITLE
Propagate authentication method claims in openid token

### ISSUE THREAD

**Issue by gberche-orange** [cloudfoundry/uaa#374]
It is sometimes required by applications to be able to know whether a user authenticated using multi factor authentication, or regular password flow. Following related discussion during the cfsummit uaa office hours, I'm submitting this issue to pursue discussion around this topic.

The ODIC specifies into http://openid.net/specs/openid-connect-core-1_0.html#IDToken the optional 

```
amr
    OPTIONAL. Authentication Methods References. JSON array of strings that are identifiers for authentication methods 
   used in the authentication. For instance, values might indicate that both password and OTP authentication methods were used. 
   The definition of particular values to be used in the amr Claim is beyond the scope of this specification. 
   Parties using this claim will need to agree upon the meanings of the values used, which may be context-specific. 
   The amr value is an array of case sensitive strings. 
```
Additionally, http://self-issued.info/docs/draft-jones-oauth-amr-values-04.html proposes some further normalized value.

In case the IDP provided normalized authentication method (e.g. a SAML IDP in SAML assertions, or an OIDC IDP in a idtoken), could this be mapped (if a translation is needed) and propagated into the openid id token ?

As a side question, in case the user identity (and its authentication method) needs to be propagated through a chain of webservices backends, using http://docs.cloudfoundry.org/api/uaa/#api-flow to exchange tokens (swapping clientid and possibly reducing scopes), should the ```amr``` be added to the access token (as a custom non standard extension) ?

**Comment by sreetummidi** [comment]
@gberche-orange Thank you creating the issue.
I will get this prioritized.

**Comment by gberche-orange** [comment]
@sreetummidi thanks for implementation of the related story, and filling the amr field in the id token

One question though, how would the AMR be propagated to a chain of backend when exchanging tokens through the api flow http://docs.cloudfoundry.org/api/uaa/#api-flow which expects a token bearer as authentication (and no an id token if I'm understanding this correctly) ?

**Comment by sreetummidi** [comment]
@gberche-orange 
If authentication was done through UAA Internal Auth then we set the AMR to pwd
If the authentication was done thru an external SAML provider we ACR instead of AMR as SAML passed Authentication Context Reference value
If the authentication was done thru an external OIDC provider we set the AMR to be the same as set by the external OIDC provider

**Comment by sreetummidi** [comment]
This story has been picked up and will be available as part of UAA 3.8.0
